Camera interface esp32s3

epc-ake opened the pull request:
# Pull Request Summary
This pull request introduces video capturing functionality for the ESP32-S3 LCD_CAM interface.  
~~To use this feature, you must first merge pull request #74974.~~ 
It depends on the [hal espressif](https://github.com/zephyrproject-rtos/hal_espressif/pull/316) being merged and added to the zephyr manifest. 
Alternatively drop the last commit so the driver writes directly into the registers and skipping the hal. 
Using the xiao_esp32s3 board and the OV2640 camera sensor, I successfully captured low resolution images in RGB565 format.

# Open Topics for Discussion
Although the driver is functional, several topics need further discussion:

## 1. Using External Memory
~~Due to issues with memory reallocation on the ESP32-S3, testing was limited to internal memory, requiring the imager to be set to the lowest resolution (160x120px).~~ 
#76982 enables to choose the video buffer location and to put it into to external ram. This enables this driver also to capture high resolution images.

## 2. ESP32-HAL
~~Currently, the driver directly writes to the LCD_CAM register values using the SOC register structs.  
It may be more appropriate to move these calls to the hal_espressif.~~ 

## 3. Devicetree structure:
The LCD_CAM module involves four different devices (cam-clk, lcd-clk, lcd-interface, cam-interface). 
Initially, I attempted to combine the cam-clk and cam-interface into one device and set the cam-clk rate based on a devicetree property. 
This approach failed because the cam-interface references the sensor (OV2640) in the device tree, requiring the sensor to be initialized before the cam-interface, which isn't possible without first initializing the camera clock.

A proposed structure is:

```
lcd_cam: lcd_cam@60041000 {
    compatible = "espressif,esp32-lcd_cam";
    reg = <0x60041000 DT_SIZE_K(4)>;
    status = "disabled";
    clocks = <&rtc ESP32_LCD_CAM_MODULE>;

    cam-xclk: cam-xclk {
        compatible = "espressif,esp32-cam-xclk";
        clock-rate = <1000000>;
    };

    cam-interface: cam-interface {
        compatible = "espressif,esp32-cam";
        dmas = <&dma 2>;
        dma-names = "rx";
    };

    lcd-clk: lcd-clk {
        compatible = "espressif,esp32-lcd-clk";
        clock-rate = <1000000>;
    };

    lcd-interface: lcd-interface {
        compatible = "espressif,esp32-lcd";
        dmas = <&dma 2>;
        dma-names = "tx";
    };
};

```

However, this might be overly complex, especially since espressif,esp32-lcd_cam would only enable the peripheral without much additional functionality.  
If you have a better idea or see an alternative way to implement this, please leave a comment!

## 4. Testing
~~I have some tests locally that only run on the xiao_esp32s3 with the ov2640 sensor. I could add them to the zephyr test folder but I don't know where to place them exactly. 
It would also be nice to add support for the `sample/subsys/video/` samples.~~ 
This PR adds an overlay for the `xiao_esp32s3` to the `samples/drivers/video/capture/`.

epc-ake committed:
drivers: video: esp32s3: add support for cam interface

Adding support for the esp32s3 camera interface.

Signed-off-by: Armin Kessler <ake@espros.com>

epc-ake committed:
drivers: video: clk: esp32s3: add camera clock device

This commit creates a new node for the lcd_cam interface.
It enables the lcd_cam pheripheral clock and adds the possibiliy
to controll the camera clock frequency trough the `cam-clk` property.
This is its own device for two reasons:
1. It enables to use the camera clock as a standalone clock.
2. It enables to initialize xclk before the sensor driver.

Signed-off-by: Armin Kessler <ake@espros.com>

epc-ake committed:
drivers: video: esp32s3: remove VSync interrupt

Remove the VSync interrupt handle, as DMA captures
VSync through the in_suc_eof event.

Signed-off-by: Armin Kessler ake@espros.com

josuah commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -0,0 +1,425 @@
+#include <zephyr/pm/device.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(esp32_cam, LOG_LEVEL_INF);
```
Currently there is no dedicated log level for video drivers, so setting the log level directly there sounds fine, as it is the role for another pull requests (i.e. that I'll do) to fix this for all video drivers.

epc-ake replied:
I would appreciate that to!

josuah commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -0,0 +1,425 @@
+	struct dma_block_config dma_blocks[CONFIG_DMA_ESP32_DESCRIPTOR_NUM];
+};
+
+struct video_esp32_dma_data {
+	struct k_fifo *fifo_out;
+	struct video_buffer *vbuf;
+};
```
Sorry if I missed it, but is that used anywhere? It looks like `struct video_esp32_data` replaced it.

epc-ake replied:
Thanks, that was a left over...

josuah commented on dts/bindings/clock/espressif,esp32-cam-clk.yaml:
```diff
@@ -0,0 +1,21 @@
+#
+# Copyright (c) 2019, Linaro Limited
```
Given this is a new driver, maybe preferable to put your own copyright?

josuah commented on dts/bindings/clock/espressif,esp32-cam-clk.yaml:
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: NXP MCUX CMOS sensor interface
```
Some remnant of the "NXP" driver here.

josuah commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -0,0 +1,425 @@
+	.set_signal = NULL,
+};
+
+PINCTRL_DT_INST_DEFINE(0);
+
+static const struct video_esp32_config esp32_config = {
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
+	.sensor_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0, sensor)),
+	.dma_dev = ESP32_DT_INST_DMA_CTLR(0, rx),
+	.rx_dma_channel = DT_INST_DMAS_CELL_BY_NAME(0, rx, channel),
+	.enable_16_bit = DT_INST_PROP(0, enable_16bit_mode),
+	.invert_bit_order = DT_INST_PROP(0, invert_bit_order),
+	.invert_byte_order = DT_INST_PROP(0, invert_byte_order),
+	.invert_pclk = DT_INST_PROP(0, invert_pclk),
+	.invert_de = DT_INST_PROP(0, invert_de),
+	.invert_hsync = DT_INST_PROP(0, invert_hsync),
+	.invert_vsync = DT_INST_PROP(0, invert_vsync)};
+
+static struct video_esp32_data esp32_data = {0};
+
+DEVICE_DT_INST_DEFINE(0, &esp32_init, PM_DEVICE_DT_INST_GET(idx), &esp32_data, &esp32_config,
+		      POST_KERNEL, CONFIG_VIDEO_INIT_PRIORITY, &esp32_driver_api);
```
Out of curiosity, are there any ESP32 with this driver and several instance of this peripheral?
If so, this could eventually be put under a `DT_INST_FOREACH_STATUS_OKAY(VIDEO_ESP32_DEVICE_DEFINE)` macro if feeling like doing this extra step.

epc-ake replied:
As far as I know. The esp32s3 is the only MCU jet that implements this peripheral and it implements only one instance. 
Therefore I keep it like this for now.

josuah replied:
(sorry for the long delay) Thank you good to note about it. This makes a lot of sense.

josuah commented on dts/bindings/video/espressif,esp32_cam.yaml:
```diff
@@ -0,0 +1,51 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: NXP MCUX CMOS sensor interface
```
Some remnant of the "NXP" driver here.

josuah commented on dts/bindings/video/espressif,esp32_cam.yaml:
```diff
@@ -0,0 +1,51 @@
+  pinctrl-names:
+    required: true
+
+  sensor:
```
I think this is name `source` now, but some drivers might still have it named `sensor` and need to be adapted...
https://github.com/zephyrproject-rtos/zephyr/pull/72623

josuah commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -0,0 +1,425 @@
+		return;
+	}
+
+	k_fifo_put(&data->fifo_out, data->active_vbuf);
+	data->active_vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
```
Interesting way to do it! With two FIFO, this is now completely asynchronous...

epc-ake replied:
Thanks. I think I got the inspiration from `video_stm32_dcmi` driver :-)

josuah commented on drivers/video/Kconfig.esp32_dvp:
```diff
@@ -0,0 +1,6 @@
+config VIDEO_ESP32S3
+	bool "Blink device drivers"
```
Likely some comment to change slightly.

josuah commented on drivers/video/Kconfig.esp32_dvp:
```diff
@@ -0,0 +1,6 @@
+	default n
+	depends on DMA
+	help
+	  This option enables the blink custom driver class.
```
Likely some comment to change slightly.

josuah reviewed (changes requested):
Thank you for this driver! I tried to spot a few things that I thought you'd be interested in adjusting. Glad to see more device families part of the Video driver collection.

ASerbinski commented:
I'm no expert with cameras or the differences between successive generations of esp32's, but is this fundamentally bound to the "s3" version of esp32? Or would it also be compatible with older esp32's as well?

epc-ake commented:
@ASerbinski The s3 is the only esp32 model with a dedicated video interface. The other esp32 variants use the I2S interface to "emulate" a video interface. Therefore, this driver is only compatible with the esp32s3.

havoc-dlt commented:
Concerning the testing, I can't seem to find your sample code anywhere. I'd really like to contribute with testing, so can you point me in the right direction?

josuah commented:
> <@epc-ake> I could add them to the zephyr test folder but I don't know where to place them exactly

The best could be to have the upstream examples supported, making modifications to the example only if needed, and using a [shield](https://docs.zephyrproject.org/latest/hardware/porting/shields.html), a bit like how it was done [for this OV5640 module](https://github.com/zephyrproject-rtos/zephyr/pull/72434/files#diff-d96c606b50f85dd1dd49593555c4ad298be15ca84bc2be3bfd1aaba39f093d05).

Maybe one way to figure out how to compose the example together would be to add another example [here](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/video) on this PR, just for the sake of publishing the source and reviewing it, even if it is removed afterward... Or just paste the snippets of code/config if there are just a few.

Thanks again!

epc-ake committed:
sample: video: esp32s3: add support for xiao_esp32s3

adding xiao_esp32s3 support to video capture sample

Signed-off-by: Armin Kessler ake@espros.com

epc-ake commented:
> Concerning the tests/samples, I can't seem to find your sample code anywhere. I'd really like to contribute with testing, so can you point me in the right direction?

I made a new commit and added the overlays for the `xiao_esp32s3` to support for the `sample/subsys/video/capture`. 
I made some temporary changes to the sample it self, since I still didn't figure out how to store the image in external ram. 
This should enable you to capture images on the ov2640 and store the image in internal ram. 
Remember that you also have to checkout #74974 for this PR to work.

havoc-dlt commented:
> > Concerning the tests/samples, I can't seem to find your sample code anywhere. I'd really like to contribute with testing, so can you point me in the right direction?
> 
> I made a new commit and added the overlays for the `xiao_esp32s3` to support for the `sample/subsys/video/capture`. I made some temporary changes to the sample it self, since I still didn't figure out how to store the image in external ram. This should enable you to capture images on the ov2640 and store the image in internal ram. Remember that you also have to checkout #74974 for this PR to work.

Thanks for the reminder, but I already have a working setup with your PRs patched in and been tinkering with it and my own zephyr projects a bit. I'm looking forward to test your code with it!

uLipe commented on drivers/clock_control/clock_control_esp32_cam.c:
```diff
@@ -0,0 +1,99 @@
+#include <soc/lcd_cam_struct.h>
+#include <soc/lcd_cam_reg.h>
+
+#define ESP32_CAM_PLL_F160M     160000000UL
```
I think these symbols could be placed into the camera device binding and fetched via dts property

epc-ake replied:
This is a fixed constant. I don't see much value adding it to the dts since it would not bee allowed to change. 
Correct me if I'm wrong.

uLipe commented on drivers/clock_control/clock_control_esp32_cam.c:
```diff
@@ -0,0 +1,99 @@
+		return ret;
+	}
+
+	printk("cam clock initialized\n");
```
Please remove or replace the printk with LOG_DBG.

uLipe commented on dts/bindings/clock/espressif,esp32-cam-clk.yaml:
```diff
@@ -0,0 +1,21 @@
+
+include: [base.yaml, pinctrl-device.yaml]
+
+properties:
```
As I mentioned above, some of the definitions hardcoded in the driver could assume a property here.

epc-ake replied:
Same as above.

uLipe commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -0,0 +1,463 @@
+LOG_MODULE_REGISTER(esp32_cam, LOG_LEVEL_INF);
+
+#define VIDEO_ESP32_F160M               160000000UL
+#define VIDEO_ESP32_DMA_BUFFER_MAX_SIZE 4095
```
These symbols can be properties in the dts bindings and selected per-project via dts/overlay as well.

epc-ake replied:
Same as above.

uLipe commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -0,0 +1,463 @@
+	}
+
+	if (dma_block_iter->next_block) {
+		LOG_ERR("Not enough descriptors available. Increase DMA_ESP32_DESCRIPTOR_NUM");
```
I noticed there is neither irq_unlock, not goto_unlock which may cause irq state conflict.

epc-ake replied:
good spot

uLipe commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -0,0 +1,463 @@
+	LCD_CAM.cam_ctrl.cam_update = 0;
+	LCD_CAM.cam_ctrl1.cam_start = 1;
+
+	data->is_streaming = true;
```
Question, is it a must to place a giant lock inside of this function?

epc-ake replied:
probably not, since the interrupts are inactive when you start the stream any how...
I'll remove it and add a check to make sure we are not already streaming on stream_start().

uLipe reviewed (changes requested):
Great work!  Please check some comments and minor concerns I placed across this patchset

epc-ake committed:
sample: video: esp32s3: implement hal_api

The driver now calls the `<hal/cam_hal.h>` API instead of writing
directly into the registers.

Signed-off-by: Armin Kessler ake@espros.com

epc-ake commented:
@LucasTambor I've added a new commit to so that the driver now uses the [espressif hal](https://github.com/zephyrproject-rtos/hal_espressif/pull/316) instead of writing directly into the registers.

kartben commented on samples/drivers/video/capture/boards/xiao_esp32s3_procpu.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_VIDEO_ESP32S3=y
+CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=40000
+CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=3
+CONFIG_CLOCK_CONTROL_ESP32_CAM=y
```
Drop. Not needed when you will have fixed the kconfigs

kartben commented on samples/drivers/video/capture/boards/xiao_esp32s3_procpu.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_DMA=y
+CONFIG_VIDEO_ESP32S3=y
```
Same

kartben commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -144,6 +144,20 @@ int main(void)
 	}
 #endif
 
+        fmt.width = 160;
+	fmt.height = 120;
+	fmt.pitch = fmt.width * 2;
+	fmt.pixelformat = VIDEO_PIX_FMT_RGB565;
+
+	if(video_set_format(video_dev, VIDEO_EP_OUT, &fmt)){
+                LOG_ERR("Unable to set format");
+                return 0;
+        }
+
+        printk("- Set format: %c%c%c%c %ux%u\n", (char)fmt.pixelformat,
+	       (char)(fmt.pixelformat >> 8), (char)(fmt.pixelformat >> 16),
+	       (char)(fmt.pixelformat >> 24), fmt.width, fmt.height);
+
```
What is this change?

epc-ake replied:
This is just a workaround to be able to fit the captured frame into internal ram. 
It should and will be removed once #76982 is merged.

kartben replied:
Ah I see. But since this impacts (negatively) users using other drivers, I would not consider this workaround acceptable. If you wanted, I guess you could add Kconfigs to the sample to allow setting a max resolution though, and based on the supported resolutions you get from the call to video_get_capabilities you could set to whatever resolutions is the closest?

LucasTambor commented on drivers/clock_control/clock_control_esp32_cam.c:
```diff
@@ -0,0 +1,99 @@
+	int ret = 0;
+
+	if (0 == cam_clk) {
+		LCD_CAM.cam_ctrl.cam_clk_sel =
```
Can you use HAL API here as well?

epc-ake replied:
I left a comment in the [hal/cam PR. ](https://github.com/zephyrproject-rtos/hal_espressif/pull/316#discussion_r1718604142)

LucasTambor commented on drivers/video/Kconfig.esp32_dvp:
```diff
@@ -0,0 +1,6 @@
+config VIDEO_ESP32S3
```
Let's use just `VIDEO_ESP32`, it's more generic for later SoCs

epc-ake commented:
Thanks for all the reviews. 
I will be on vacation for the next two weeks and won't be able to work on this. 
If you need to expedite it, feel free to create a new PR based on this one. 
Otherwise, I will implement your suggestions after I return.

epc-ake commented:
Since @LucasTambor took over this at #77709 and #77879 I close this PR
